jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -719,4 +727,23 @@ public static String buildPersistenceUnitName(URL url, String puName){
        return fullPuName;
    }
 
+    /**
+     * @param file archive file URL
+     * @param entry a directory entry in the archive (or an empty string)
+     * @return true if the file-entry pair can be a persistence root according
+     * to JPA Spec (8.2).
+     */
+    private static boolean isValidRootInArchive(String file, String rootEntry) {
+        String extension = file.substring(Math.max(0, file.length() - 4));
+        boolean valid;
+        if (extension.equalsIgnoreCase(".jar")) {
+            // For a JAR, the root can only be the archive itself.
+            return rootEntry.isEmpty();
+        } else if (extension.equalsIgnoreCase(".war")) {
+            // For a WAR, the root can only be WEB-INF/classes.
```

Comment:
:/ Not exactly true... 

Spec 8.2: 

    a jar file in the WEB-INF/lib directory of a WAR file

Reply:
Hm, the comment is indeed a tad misleading. Any idea how I should rephrase this? Maybe, just copy&paste examples from the spec?

Reply:
Ya, I think that should be fine. I think it would be good to add the same comments to the matching tests with the spec quotes.

Reply:
Just checking back up on this; sorry, Ive been busy. Were you planning on adding the ```WEB-INF/lib``` case as valid for ```isValidRootInArchive()```?

Reply:
This gets kind of weird. The spec says:
```
a jar file in the WEB-INF/lib directory of a WAR file
```
which means that `pxmlURL` in this case must be an entry inside of an entry. I know that this isn't supported in the `jar:` URLs. Could you give me an example of such URLs for other schemes?

In any case, I'm not sure it makes sense to forbid all such nesting archives except the ones explicitly permitted by the spec.

Reply:
I'm guessing this would be an exploded war? So I think the URL would look like this? : ```jar:file:/foo/bar.war/WEB-INF/lib/jpa.jar!``` I'm actually not sure. Perhaps this isn't a special case for ```war``` in this method.

Reply:
> I'm guessing this might come in as an exploded war?

For exploded war we can't really validate this. If the file path ends in `WEB-INF/lib/foo.jar` - it might be a lib JAR inside an exploded WAR, if not - it's just a stand-alone JAR. Both are possible, and both should be allowed, so it doesn't matter to check in this case.

The problem is what happens if the WAR is NOT exploded. At least Tomcat seems to [allow that](https://tomcat.apache.org/tomcat-8.5-doc/config/context.html) (search for `unpackWar`), though I don't know the details.

> My default is just that we should cover all the spec situations.

That is the easy part. We take the URL to the descriptor resource, strip the descriptor location postfix from it - and that's it. Some complexity arises from nuances of packaging and file system differences, but that's already done. We're now talking about the **validation** for possible situations (raising an exception if we detect something that isn't *explicitly* allowed by the spec). This is a bit of a feature creep for this PR, IMO.

> I'm open to interruption + it helps if we can see how the previous code would have worked. If the old code didnt cover this case, however we decide that looks, then at least we don't have a regression :/

1. JAR URLs don't allow nested archives anyway.
2. For ZIP/WSJAR URLs we (erroneously) used to strip the entry entirely, whereas now we just strip the `META-INF/persistence.xml` postfix (which should work . This is the point of this PR.

I guess the prudent thing to do here would be to check what URLs we get from different servers for nested archives, and **validate** based on that. However, this looks like a lot of work, and this PR has already taken far more time than I consider reasonable for this (seemingly trivial) fix. Can I limit its scope to just fixing the algorithm bug, and leave adding the validation (except the descriptor location check) for some other time?

Reply:
If we can skip the "jar inside war" situation, then I think your change is good as is and I'll approve it. Apologies for having this bug go further than you intended. I want to make sure this is fixed in a way that is good for all situations and that the testing is correct.

Reply:
Ok, I've clarified the assumptions this method does about its arguments, and expanded the comment in question.